File: 413.等差数列划分.go
/*
 * @lc app=leetcode.cn id=413 lang=golang
 *
 * [413] 等差数列划分
 */

// @lc code=start
func numberOfArithmeticSlices(nums []int) (res int) {
    if len(nums) < 3 {
        return
    }
    d, t := nums[0] - nums[1], 0
    for i := 2; i < len(nums); i++ {
        if nums[i - 1] - nums[i] == d {
            t++
        } else {
            d = nums[i - 1] - nums[i]
            t = 0
        }
        res += t
    }
    return
}
// @lc code=end

File: 413.等差数列划分.py
#
# @lc app=leetcode.cn id=413 lang=python3
#
# [413] 等差数列划分
#

# @lc code=start
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return 0

        d, t = nums[0] - nums[1], 0
        res = 0

        # 因为等差数列的长度至少为 3，所以可以从 i=2 开始枚举
        for i in range(2, n):
            if nums[i - 1] - nums[i] == d:
                t += 1
            else:
                d = nums[i - 1] - nums[i]
                t = 0
            res += t

        return res
# @lc code=end
